### PR TITLE
Pass through logs for 0 confirmations

### PR DESCRIPTION
**Pull request by PiotrTrzpil**
If all subscribers are setup with 0 confirmations, do not check logs' block numbers when sending them

### CODE REVIEW

File: core/services/log/broadcaster.go [hunk 1]
```diff
@@ -401,18 +405,32 @@ func (b *broadcaster) onNewHeads() {
...
-		logs, minBlockNum := b.logPool.getLogsToSend(latestBlockNum)
-
-		if len(logs) > 0 {
-			broadcasts, err := b.orm.FindConsumedLogs(minBlockNum, latestBlockNum)
-			if err != nil {
-				logger.Errorf("LogBroadcaster: Failed to query for log broadcasts, %v", err)
-				return
+		// if all subscribers requested 0 confirmations, we always get and delete all logs from the pool,
+		// without comparing their block numbers to the current head's block number.
+		if b.registrations.lowestNumConfirmations == 0 && b.registrations.highestNumConfirmations == 0 {
```
Comment (connorwstein) [inline core/services/log/broadcaster.go:410]: Wouldn't just b.registrations.highestNumConfirmations == 0 be enough here?

File: core/services/log/broadcaster_test.go [hunk 2]
```diff
@@ -520,6 +520,93 @@ func TestBroadcaster_BroadcastsAtCorrectHeights(t *testing.T) {
...
+	})
+	defer cleanup()
+
+	chRawLogs := <-helper.chchRawLogs
+
+	for _, log := range addr1SentLogs {
+		chRawLogs <- log
+	}
+
+	requireBroadcastCount(t, helper.store, 6)
```
Comment (connorwstein) [inline core/services/log/broadcaster_test.go:563]: I don't love this requireBroadcastCount as it leans on gomegas consistently + eventually. Could we just install a buffered channel in the helper which HandleLog writes to and we just read from 6 times here (or timeout) ?

File: core/services/log/pool.go [hunk 3]
```diff
@@ -29,6 +31,39 @@ func (pool *logPool) addLog(log types.Log) {
...
+			break
+		}
+
+		blockNum := uint64(item.(Uint64))
+		logs, exists := pool.hashesByBlockNumbers[blockNum]
+		if exists {
+			if int64(blockNum) < lowest {
+				lowest = int64(blockNum)
+			}
+			if int64(blockNum) > highest {
```
Comment (samsondav) [inline core/services/log/pool.go:51]:
```suggestion
			} else if int64(blockNum) > highest {
```

File: core/services/log/registrations.go [hunk 4]
```diff
@@ -61,47 +63,50 @@ func newRegistrations() *registrations {
 	return &registrations{
 		subscribers: make(map[uint64]*subscribers),
 		decoders:    make(map[common.Address]ParseLogFunc),
+
+		// a high value for lowestNumConfirmations is set until first subscriber,
+		// to avoid a special case in comparison to its NumConfirmations
+		lowestNumConfirmations: uint64(math.MaxUint64),
```
Comment (samsondav) [inline core/services/log/registrations.go:69]:
```suggestion
		lowestNumConfirmations: math.MaxUint64,
```

File: core/services/log/registrations.go [hunk 5]
```diff
@@ -61,47 +63,50 @@ func newRegistrations() *registrations {
...
+	if len(r.subscribers[reg.opts.NumConfirmations].handlers) == 0 {
+		delete(r.subscribers, reg.opts.NumConfirmations)
+		r.resetNumConfirmationsRange()
 	}
 	return
 }
 
-// increase the highestNumConfirmations stored if the new listener has a higher value
-func (r *registrations) maybeIncreaseHighestNumConfirmations(newNumConfirmations uint64) {
+// maybe update the numbers tracking highest and lowest num confirmations
```
Comment (samsondav) [inline core/services/log/registrations.go:102]: we can do better than this as a comment

File: core/services/log/registrations.go (hunk 4 above)
Comment (samsondav) [reply core/services/log/registrations.go:69]: Alternatively use an `int64` and `-1` as the sentinel value, or even a nullable pointer

**Review by samsondav — CHANGES_REQUESTED** [review]

File: core/services/log/broadcaster.go (hunk 1 above)
Comment (PiotrTrzpil) [reply core/services/log/broadcaster.go:410]: ah yes actually very good point!

File: core/services/log/registrations.go (hunk 4 above)
Comment (PiotrTrzpil) [reply core/services/log/registrations.go:69]: removed this field now, it wasn't actually needed

File: core/services/log/registrations.go (hunk 5 above)
Comment (PiotrTrzpil) [reply core/services/log/registrations.go:102]: updated, and moved out of the separate method

File: core/services/log/pool.go (hunk 3 above)
Comment (PiotrTrzpil) [reply core/services/log/pool.go:51]: hm no, it's not ok to have an `else if` - both can be true (usually will) at the same time on the first iteration

File: core/services/log/broadcaster_test.go (hunk 2 above)
Comment (PiotrTrzpil) [reply core/services/log/broadcaster_test.go:563]: do we want to reduce the usage of gomega? i thought it's quite handy - e.g. compared to `require.Eventually` (e.g. `require.Eventually(t, func() bool { return len(idleDurationOccured) == 2 }, 3*time.Second, 10*time.Millisecond)` ) it prints the actual value at the timeout, if it's as expected

Comment (PiotrTrzpil) [reply core/services/log/broadcaster_test.go:563]: but yes a more direct check of sends will be nice, rather than going via the DB

Comment (PiotrTrzpil) [reply core/services/log/broadcaster_test.go:563]: hm though it's a bit tricky - currently simpleLogListener does not have a reference to the helper (where i'd put the channel) - so i'm rather tempted to do it in another PR

**Review by samsondav — APPROVED** [review]

**Review by connorwstein — APPROVED** [review]

**Pull request merged**